django/contrib/admin/options.py

```diff
@@ -1929,6 +1929,9 @@ def _create_formsets(self, request, obj, change):
         if change:
             get_formsets_args.append(obj)
         for FormSet, inline in self.get_formsets_with_inlines(*get_formsets_args):
+            # Do not create inline for view-only user
+            if not (inline._has_add_permission(request, obj) or inline.has_change_permission(request, obj)):
+                continue
```

Comment:
This needs to be filtered behind a `request.method == 'POST'`. But I want to add extra tests here, since nothing fails, even though the behaviour is wrong.

Reply:
Should it consider "has_delete_permission" also? Maybe a method that encompasses this condition would help readability and ultimately be used elsewhere.

Reply:
I tried a similar approach before opening a (duplicate) ticket, as well as one related to issues when POST fails. I ran into the error that forms have been messed with.

If POST fails, the output should be the same as in GET, with errors, of course. Everything else is a hack, IMHO. The only solution that seemed a good one to me is to deal with this at the point of validation, but that code is not in admin, so I concluded this was something for the core team.

Reply:
Thanks for the comment @matijakolaric: I do agree with you. I'm looking at this again later today. First it's putting the extra test cases in, then we'll work out the best solution.